src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -581,6 +618,142 @@ fn max_slice_length<'p, 'a: 'p, 'tcx: 'a, I>(
     cmp::max(max_fixed_len + 1, max_prefix_len + max_suffix_len)
 }
 
+/// An inclusive interval, used for precise integer exhaustiveness checking.
+/// `Interval`s always store a contiguous range of integers. This means that
+/// signed values are encoded by offsetting them such that `0` represents the
+/// minimum value for the integer, regardless of sign.
+/// For example, the range `-128...127` is encoded as `0...255`.
+/// This makes comparisons and arithmetic on interval endpoints much more
+/// straightforward. See `offset_sign` for the conversion technique.
+struct Interval<'tcx> {
+    pub range: RangeInclusive<u128>,
+    pub ty: Ty<'tcx>,
+}
+
+impl<'tcx> Interval<'tcx> {
+    fn from_ctor(tcx: TyCtxt<'_, 'tcx, 'tcx>,
+                 ctor: &Constructor<'tcx>)
+                 -> Option<Interval<'tcx>> {
+        match ctor {
+            ConstantRange(lo, hi, end) => {
+                assert_eq!(lo.ty, hi.ty);
+                let ty = lo.ty;
+                if let Some(lo) = lo.assert_bits(ty) {
+                    if let Some(hi) = hi.assert_bits(ty) {
+                        // Perform a shift if the underlying types are signed,
+                        // which makes the interval arithmetic simpler.
+                        let (lo, hi) = Self::offset_sign(tcx, ty, lo..=hi, true);
+                        // Make sure the interval is well-formed.
+                        return if lo > hi || lo == hi && *end == RangeEnd::Excluded {
+                            None
+                        } else {
+                            let offset = (*end == RangeEnd::Excluded) as u128;
+                            Some(Interval { range: lo..=(hi - offset), ty })
+                        };
+                    }
+                }
+                None
+            }
+            ConstantValue(val) => {
+                let ty = val.ty;
+                if let Some(val) = val.assert_bits(ty) {
+                    let (lo, hi) = Self::offset_sign(tcx, ty, val..=val, true);
+                    Some(Interval { range: lo..=hi, ty })
+                } else {
+                    None
+                }
+            }
+            Single | Variant(_) | Slice(_) => {
+                None
+            }
+        }
+    }
+
+    fn offset_sign(tcx: TyCtxt<'_, 'tcx, 'tcx>,
+                   ty: Ty<'tcx>,
+                   range: RangeInclusive<u128>,
+                   encode: bool)
```

Comment:
Maybe have encode/decode methods? Which could be named `from_bits` and `to_bits`. Or just `encode` / `decode`.